# Exercicios

1. Uma moeda está viciada: a chance de se obter uma cara usando ela é de 0,6. Qual probabilidade de essa moeda apresentar, em qualquer ordem, 3 caras e 7 coroas.

In [89]:
from itertools import permutations, combinations, combinations_with_replacement, product

espaco = product(['C', 'K'], repeat=10)

prob_cara = 0.6
prob_coroa = 1 - prob_cara
prob_total = 0

for evento in espaco:
    n_caras = evento.count('C')
    n_coroas = evento.count('K')
    
    if n_caras == 3 and n_coroas == 7:
        prob_total += prob_cara**n_caras * prob_coroa**n_coroas
    
print(prob_total)

0.04246732799999992


2. Como reproduzir o resultado acima usando os conceitos de permutação, arranjo, etc....?


> Se designarmos por A o evento que consiste na obtenção 3 caras e 7 coroas em 10 lançamentos, então:
$$P(A) = 0.6^3 * 0.4^7 * P_{10}^{7,3} $$

$$P_{10}^{7,3} = \frac{10!}{7!3!} $$

$$P(A) = 0.6^3 * 0.4^7 * \frac{10!}{7!3!} $$

$$P(A) = 0.04246733 $$

In [87]:
def fatorial (numero):
    resultado_fatorial = 1
    
    for i in range(1, numero + 1):
        if numero > 0:
            resultado_fatorial *= i

    return resultado_fatorial


P = (0.6**3 * 0.4**7) * fatorial(10)/(fatorial(7)*fatorial(3))

0.04246732800000001


3. Em um programa de auditório, havia 3 portas (A, B, C), dentro da qual uma continha um prêmio. O participante escolheu a porta A e antes de abrir a porta o apresentador abriu a porta C, na qual não havia nada. O participante aumentará a probabilidade de acertar a porta caso mude a escolha dele da porta A para a B?

Importando bibliotecas necessárias

In [10]:
import random

Probabilidade de sucesso sem mudança de portas

In [81]:
portas = ['A', 'B', 'C']

qnt_tentativas = 100000
sucessos = 0

for i in range(qnt_tentativas):
    porta_premiada = portas[random.randint(0,2)]
    porta_escolhida = portas[random.randint(0,2)]

    if porta_premiada == porta_escolhida:
        sucessos += 1

probabilidade_sucesso = sucessos/qnt_tentativas
print(f'A probabilidade de sucesso estimada para essa estratégia é de: {probabilidade_sucesso*100:.2f}%')

A probabilidade de sucesso estimada para essa estratégia é de: 33.36%


Probabilidade de sucesso com mudança de portas:

In [72]:
qnt_tentativas = 100000
sucessos = 0

for i in range(qnt_tentativas):
    portas = ['A', 'B', 'C']
    
    porta_premiada = portas[random.randint(0,2)] #Definindo a porta premiada aleatoriamente
    porta_escolhida = portas[random.randint(0,2)] #Definindo a porta escolhida aleatoriamente
    porta_aberta = ''

    #Abrindo uma porta (não pode ser a escolhida pelo participante ou a premiada)
    while len(portas) == 3:
        porta_aberta = portas[random.randint(0,2)]

        if porta_aberta != porta_escolhida and porta_aberta != porta_premiada: #Verificando se a porta aberta é a escolhida pelo participante ou a premiada
            portas.remove(porta_aberta) #Se ela não for, removemos a porta aberta da lista de portas disponíveis. Se não o loop reinicia e escolhemos outra porta.
    
    #Trocando a porta escolhida inicialmente pela porta que sobrou
    if portas.index(porta_escolhida) == 0:
        porta_escolhida = portas[1]
    elif portas.index(porta_escolhida) == 1:
        porta_escolhida = portas [0]

    #Verificando se o participante ganhou o prêmio:
    if porta_escolhida == porta_premiada:
        sucessos += 1
    
probabilidade_sucesso = sucessos/qnt_tentativas
print(f'A probabilidade de sucesso estimada para essa estratégia é de: {probabilidade_sucesso*100:.2f}%')

A probabilidade de sucesso estimada para essa estratégia é de: 66.68%


Estimamos as probabilidade de sucesso para dois cenários: 
 - No primeiro cenário o jogador não troca de porta e carrega a sua escolha incial até o fim do jogo; 
 - No segundo cenário o jogador opta por trocar de porta quando uma das portas vazias é revelada.

Para o primeiro cenário estimamos um probabilidade média de sucesso de 33.33%, para o segundo cenário a probabilidade de sucesso aumentou para 66.66%! 

Por que isso aconteceu?

Podemos analisar esse problema sob a seguinte perspectiva: 



Como temos 3 portas e apenas uma delas é a premiada, o jogador inicialmente tem um probabilidade de <b>33% de acertar</b> na sua escolha e <b>66% de errar</b>.

Se o jogador escolheu a porta premiada inicialmente (33% de chance, como mencionado anteriormente), o apresentador vai abrir uma das portas vazias e vai oferecer a oportunidade de trocar a sua porta inicial (premiada) pela porta que sobrou (uma das portas vazias). Caso você opte por trocar, você vai terminar com a porta vazia.

Agora, se o jogador escolheu inicialmente uma das portas vazias (66% de chance), o apresentador vai abrir a outra porta vazia e vai oferecer a oportunidade de trocar a sua porta incial (vazia) pela porta que sobrou (premiada). Caso você opte trocar você vai terminar com a porta premiada.

Então, se a sua estratégia for sempre trocar de porta quando oferecido, você terá as suas chances invertidas: caso você inicialmente escolha inicialmente uma das portas vazias, ao trocar você terminará com a porta premiada (<b>66% de chance de ganhar</b>); e caso você escolha inicialmente a porta premiada, ao trocar
 você terminará com uma porta vazia (<b>33% de chance de perder</b>)




